tests/Mono.Android-Tests/Xamarin.Android.Net/AndroidMessageHandlerTests.cs

```diff
@@ -19,6 +20,48 @@ protected override HttpMessageHandler CreateHandler ()
 			return new AndroidMessageHandler ();
 		}
 
+		// We can't test `deflate` for now because it's broken in the BCL for https://httpbin.org/deflate (S.I.Compression.DeflateStream doesn't recognize the compression
+		// method used by the server)
+		static readonly object[] DecompressionSource = new object[] {
+			new object[] {
+				"gzip", // urlPath
+				"gzip", // encoding
+				"gzipped", // jsonFieldName
+			},
+
+			new object[] {
+				"brotli", // urlPath
+				"br", // encoding
+				"brotli", // jsonFieldName
+			},
+		};
+
+#if NET
+		[Test]
+		[TestCaseSource (nameof (DecompressionSource))]
+		public async Task Decompression (string urlPath, string encoding, string jsonFieldName)
+		{
+			var handler = new AndroidMessageHandler {
+				AutomaticDecompression = DecompressionMethods.All
+			};
+
+			var client = new HttpClient (handler);
+			HttpResponseMessage response = await client.GetAsync ($"https://httpbin.org/{urlPath}");
+
+			foreach (string enc in response.Content.Headers.ContentEncoding) {
+				if (String.Compare (enc, encoding, StringComparison.Ordinal) == 0) {
+					Assert.Fail ($"Encoding '{encoding}' should have been removed from the Content-Encoding header");
+				}
+			}
+
+			string responseBody = await response.Content.ReadAsStringAsync ();
+
+			Assert.IsTrue (responseBody.Length > 0, "Response was empty");
+			Assert.AreEqual (response.Content.Headers.ContentLength, responseBody.Length, "Retrieved data length is different than the one specified in the Content-Length header");
+			Assert.IsTrue (responseBody.Contains ($"\"{jsonFieldName}\"", StringComparison.OrdinalIgnoreCase), $"\"{jsonFieldName}\" should have been in the response JSON");
```

Comment:
It's strange that it's failing in some test configurations and passing in other, especially for `"gzipped"`. Maybe we could dump the `responseBody` to see if the response from the server is what we expect?

Is it possible that sometimes the server doesn't respond with a 200 (because of some rate limitting)? We should ignore the test result if the response is 4xx or 5xx.

Reply:
Yeah, I found it curious too. Rate limiting is a possibility, definitely. It would be great if we could host httpbin somewhere, probably need to ask around about this. 

I ran the tests many times locally and always got the valid answer, so the principle of the test is solid. I will add check for 4xx and 5xx errors.